src/trajectory/util.cpp

```diff
@@ -0,0 +1,314 @@
+  traj.evaluateDerivative(partialStartTime, 1, segStartVel);
+
+  // Find a segment of original trajectory that includes partialStartTime
+  while (currSegmentIdx < traj.getNumSegments())
+  {
+    currSegmentEndTime += traj.getSegmentDuration(currSegmentIdx);
+    if (currSegmentStartTime <= partialStartTime
+        && partialStartTime <= currSegmentEndTime)
+    {
+      // create new segment
+      traj.evaluate(currSegmentEndTime, segmentEndState);
+      stateSpace->logMap(segmentEndState, segEndPos);
+      traj.evaluateDerivative(currSegmentEndTime, 1, segEndVel);
+
+      const double segmentDuration = currSegmentEndTime - partialStartTime;
+
+      if (segmentDuration > 0.0)
+      {
+        CubicSplineProblem problem(Vector2d{0., segmentDuration}, 4, dimension);
+        problem.addConstantConstraint(0, 0, zeroPos);
+        problem.addConstantConstraint(0, 1, segStartVel);
+        problem.addConstantConstraint(1, 0, segEndPos - segStartPos);
+        problem.addConstantConstraint(1, 1, segEndVel);
+        const auto solution = problem.fit();
+        const auto coefficients = solution.getCoefficients().front();
+        outputTrajectory->addSegment(
+            coefficients, segmentDuration, segmentStartState);
+      }
+      break;
+    }
+
+    currSegmentIdx++;
+    currSegmentStartTime = currSegmentEndTime;
+  }
+
+  for (std::size_t i = currSegmentIdx + 1u; i < traj.getNumSegments(); i++)
+  {
+    outputTrajectory->addSegment(
+        traj.getSegmentCoefficients(i),
+        traj.getSegmentDuration(i),
+        traj.getSegmentState(i));
+  }
+
+  return outputTrajectory;
+}
+
+//==============================================================================
+UniqueInterpolatedPtr convertToInterpolated(
+    const Spline& traj, statespace::ConstInterpolatorPtr interpolator)
+{
+  auto stateSpace = traj.getStateSpace();
+  auto outputTrajectory
+      = make_unique<Interpolated>(stateSpace, std::move(interpolator));
+
+  auto state = stateSpace->createState();
+  for (std::size_t i = 0; i < traj.getNumWaypoints(); i++)
+  {
+    traj.getWaypoint(i, state);
+    const double t = traj.getWaypointTime(i);
+    outputTrajectory->addWaypoint(t, state);
+  }
+
+  return outputTrajectory;
+}
+
+//==============================================================================
+UniqueInterpolatedPtr concatenate(
+    const Interpolated& traj1, const Interpolated& traj2)
+{
+  if (traj1.getStateSpace()->getDimension()
+      != traj2.getStateSpace()->getDimension())
+    throw std::runtime_error("Dimension mismatch");
+
+  auto stateSpace = traj1.getStateSpace();
+
+  auto outputTrajectory
+      = make_unique<Interpolated>(stateSpace, traj1.getInterpolator());
+
+  for (std::size_t i = 0; i < traj1.getNumWaypoints() - 1; i++)
+  {
+    outputTrajectory->addWaypoint(
+        traj1.getWaypointTime(i), traj1.getWaypoint(i));
+  }
+  outputTrajectory->addWaypoint(traj1.getEndTime(), traj2.getWaypoint(0));
```

Comment:
Why is this outside the loop below? What's wrong with starting that loop at 0?

Reply:
I don't know why but at least it's not identical if we put this line into the loop with starting from 0 index. More importantly, this logic will not work as expected (traj1.duration + traj2.duration != newTraj.duration) when the start time of traj2 is not zero. Let me fix this and add a corresponding test for it.

Reply:
Yep, I agree that the new fix respects a nonzero start time in traj2 better!